pkg/controller/clusterroleaggregation/clusterroleaggregation_controller.go

```diff
@@ -121,17 +122,33 @@ func (c *ClusterRoleAggregationController) syncClusterRole(key string) error {
 		return nil
 	}
 
-	// we need to update
-	clusterRole := sharedClusterRole.DeepCopy()
-	clusterRole.Rules = nil
-	for _, rule := range newPolicyRules {
-		clusterRole.Rules = append(clusterRole.Rules, *rule.DeepCopy())
-	}
-	_, err = c.clusterRoleClient.ClusterRoles().Update(context.TODO(), clusterRole, metav1.UpdateOptions{})
+	clusterRole := rbacv1ac.ClusterRole(sharedClusterRole.Name).
+		WithRules(toApplyPolicyRules(newPolicyRules)...)
+
+	opts := metav1.ApplyOptions{FieldManager: "clusterrole-aggregation-controller", Force: true}
+	_, err = c.clusterRoleClient.ClusterRoles().Apply(context.TODO(), clusterRole, opts)
 
 	return err
 }
 
+func toApplyPolicyRules(rules []rbacv1.PolicyRule) []*rbacv1ac.PolicyRuleApplyConfiguration {
+	var result []*rbacv1ac.PolicyRuleApplyConfiguration
+	for _, rule := range rules {
+		result = append(result, toApplyPolicyRule(rule))
+	}
+	return result
+}
+
+func toApplyPolicyRule(rule rbacv1.PolicyRule) *rbacv1ac.PolicyRuleApplyConfiguration {
```

Comment:
seems like we could generate this into the applyconfiguration package.

Reply:
We're only lucky here because all the fields are pointers. If they weren't, we wouldn't be able to know if the zero value needs to be copied over.

Reply:
Yeah, the problem example I've been using to try to explain why we need to be careful copying from go structs is:

- Developer uses a go struct to create an HPA that changes the `MinReplicas` value, e.g.:

```
autoscalingv1.HorizontalPodAutoscaler{
   // ...
  Spec: autoscalingv1.HorizontalPodAutoscalerSpec{
    MinReplicas: 0,
  }
}
```

- Developer converts from the go struct to an apply configuration

Doesn't really matter how. They could marshal to YAML or JSON, or they use some mechanism to copy the struct into a generated apply builder.

- Developer applies to the server
- `MaxReplicas` get's applied with the value `0` (not to be confused with `MinReplicas` which gets set to `0` like requested)
- Whatever the HPA controls is autoscaled down to 0, quite plausibly causing an outage

This is because `MaxReplicas` is a non-nil int32 that does not have the `omitempty` JSON tag:

https://github.com/kubernetes/kubernetes/blob/9a9a9b014c4e47325e10de837a720633ab8a2030/pkg/apis/autoscaling/types.go#L88

There are a lot of fields like this.

Reply:
> We're only lucky here because all the fields are pointers. If they weren't, we wouldn't be able to know if the zero value needs to be copied over.

"make every field of this match" seems fairly common, but it would also be additive.